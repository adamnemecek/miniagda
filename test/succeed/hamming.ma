-- Nat ---------------------------------------------------------------

data Nat : Set 
{ zero : Nat
; succ : Nat -> Nat 
}

fun add : Nat -> Nat -> Nat 
{ add  zero    = \y -> y
; add (succ x) = \y -> succ (add x y)
}

let double : Nat -> Nat
           = \ n -> add n n
let triple : Nat -> Nat
           = \ n -> add n (double n)

fun leq : Nat -> Nat -> [C : Set] -> C -> C -> C
{ leq  zero     y       C tt ff = tt
; leq (succ x)  zero    C tt ff = ff
; leq (succ x) (succ y) C tt ff = leq x y C tt ff 
}

-- Stream ------------------------------------------------------------

sized codata Stream (+ A : Set) : Size -> Set 
{
  cons : [i : Size] -> A -> Stream A i -> Stream A ($ i)
}

cofun map : [A : Set] -> [B : Set] -> [i : Size] -> 
            (A -> B) -> Stream A i -> Stream B i 
{
  map A B ($ i) f (cons .A .i x xl) = cons B _ (f x) (map A B _ f xl)
}

cofun merge : [i : Size] -> Stream Nat i -> Stream Nat i -> Stream Nat i
{
  merge ($ i) (cons .Nat .i x xs) (cons .Nat .i y ys) = 
      leq x y (Stream Nat _)
         (cons Nat _ x (merge _ xs (cons Nat _ y ys)))
	 (cons Nat _ y (merge _ (cons Nat _ x xs) ys))     
}


-- Hamming function --------------------------------------------------

cofun ham : [i : Size] -> Stream Nat i
{
  ham ($ i) = cons Nat _ (succ zero) 
                (merge i (map Nat Nat i double (ham i)) 
                         (map Nat Nat i triple (ham i)))
}
