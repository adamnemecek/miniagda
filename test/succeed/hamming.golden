MiniAgda by Andreas Abel and Karl Mehltretter
--- opening "hamming.ma" ---
--- scope checking ---
--- type checking ---
type  Nat : Set
term  Nat.zero : < Nat.zero : Nat >
term  Nat.succ : ^(y0 : Nat) -> < Nat.succ y0 : Nat >
term  add : Nat -> Nat -> Nat
{ add Nat.zero = \ y -> y
; add (Nat.succ x) = \ y -> Nat.succ (add x y)
}
term  double : Nat -> Nat
term  double = \ n -> add n n
term  triple : Nat -> Nat
term  triple = \ n -> add n (double n)
term  leq : Nat -> Nat -> .[C : Set] -> C -> C -> C
{ leq Nat.zero y [C] tt ff = tt
; leq (Nat.succ x) Nat.zero [C] tt ff = ff
; leq (Nat.succ x) (Nat.succ y) [C] tt ff = leq x y [C] tt ff
}
type  Stream : ++(A : Set) -> - Size -> Set
term  Stream.cons : .[A : Set] -> .[i : Size] -> ^(y1 : A) -> ^(y2 : Stream A i) -> < Stream.cons i y1 y2 : Stream A $i >
term  map : .[A : Set] -> .[B : Set] -> .[i : Size] -> (A -> B) -> Stream A i -> Stream B i
{ map [A] [B] $[i < #] f (Stream.cons [.i] x xl) = Stream.cons [i] (f x) (map [A] [B] [i] f xl)
}
term  merge : .[i : Size] -> Stream Nat i -> Stream Nat i -> Stream Nat i
{ merge $[i < #] (Stream.cons [.i] x xs) (Stream.cons [.i] y ys) = leq x y [Stream Nat $i] (Stream.cons [i] x (merge [i] xs (Stream.cons [i] y ys))) (Stream.cons [i] y (merge [i] (Stream.cons [i] x xs) ys))
}
term  ham : .[i : Size] -> Stream Nat i
{ ham $[i < #] = Stream.cons [i] (Nat.succ Nat.zero) (merge [i] (map [Nat] [Nat] [i] double (ham [i])) (map [Nat] [Nat] [i] triple (ham [i])))
}
--- evaluating ---
--- closing "hamming.ma" ---
